### PR TITLE
New wrapper `FeatureShare`

### LINKED ISSUE #2099: Share Inception v3 network between FID, KID, Inception Score, and MiFID
## 🚀 Feature

A single copy of Inception v3 loaded on a GPU and shared between all the metrics which use it.

### Motivation

FID, KID, Inception Score, and MiFID all use the same Inception v3 network, which uses 304MB of GPU memory when loaded. If you want to calculate all 4 scores while training, that's 1.2GB GPU memory per GPU. It would be ideal if the metrics could share a single copy of the network to minimize GPU memory usage.

### Pitch

Torchmetrics should know if one metric which uses Inception V3 has been loaded and share the copy between the metrics, so when a second metric is created, a second copy of the network is not created.

### Alternatives

1) Documentation / officially support the "unofficial non supported" way to do this -- pass a reference to the network from one metric into the `__init__` of another metric. So for example, today you could do:

```
my_fid = torchmetrics.image.fid.FrechetInceptionDistance()
my_is = torchmetrics.image.inception.InceptionScore(feature=my_fid.inception)
```

However I am hesitant to rely on this as it depends on the internal details of metrics which are subject to change release to release. I would not know from release notes if this was broken or not.

2) A "parent metric" which created one copy of InceptionV3 and the metrics you selected to use it with, and calls to `update` / `reset` / `compuse` would in turn pass that down to each individual metric. This could get a bit messy as the `__init__` APIs do not exactly line up.

### Additional context

It would be doubly awesome if the solution allowed you to share _activations_ as well as the network itself, as currently for each copy I'm calculating the same activations on each copy of the network.

Thanks so much!

### PR DESCRIPTION
**Pull request by SkafteNicki**
## What does this PR do?

Fixes #2099
Introduces a new wrapper `FeatureShare` that is intended to be used with metrics that uses some underlying neural network to do their calculation. As pointed out in the issue it is a waste of memory that the same network that can be shared between networks gets initialized multiple times and that the computations in not reused.

Work in progress...

<details>
  <summary>Before submitting</summary>

- [ ] Was this **discussed/agreed** via a Github issue? (no need for typos and docs improvements)
- [ ] Did you read the [contributor guideline](https://github.com/Lightning-AI/torchmetrics/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure to **update the docs**?
- [ ] Did you write any new **necessary tests**?

</details>

<details>
  <summary>PR review</summary>

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

</details>

## Did you have fun?

Make sure you had fun coding 🙃

----
:books: Documentation preview :books:: https://torchmetrics--2120.org.readthedocs.build/en/2120/

### CODE REVIEW

File: src/torchmetrics/wrappers/feature_share.py [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+from torchmetrics.metric import Metric
+from torchmetrics.utilities import rank_zero_warn
+
+__doctest_requires__ = {("FeatureShare",): ["torch_fidelity"]}
+
+
+class NetworkCache(Module):
+    """Create a cached version of a network to be shared between metrics.
+
+    Because the different metrics invoke the same network multiple times, we can save time by caching the input-output
```
Comment (Borda) [inline src/torchmetrics/wrappers/feature_share.py:29]:
```suggestion
    Because the different metrics may invoke the same network multiple times, we can save time by caching the input-output
```

File: src/torchmetrics/wrappers/feature_share.py [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+        self.max_size = max_size
+        self.network = lru_cache(maxsize=self.max_size)(network)
+
+    def forward(self, *args: Any, **kwargs: Any) -> Any:
+        """Call the network with the given arguments."""
+        return self.network(*args, **kwargs)
+
+
+class FeatureShare(MetricCollection):
+    """Specialized metric collection that can be used to share features between metrics.
```
Comment (Borda) [inline src/torchmetrics/wrappers/feature_share.py:45]:
```suggestion
    """Specialized metric collection that facilitates sharing features between metrics.
```

File: src/torchmetrics/wrappers/feature_share.py [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+
+    Certain metrics rely on an underlying expensive neural network for feature extraction when computing the metric.
+    This wrapper allows to share the feature extraction between multiple metrics, which can save a lot of time and
+    memory. This is achieved by making a shared instance of the network between the metrics and secondly by caching
+    the input-output pairs of the network, such the subsequent calls to the network with the same input will be much
+    faster.
+
+    Args:
+        metrics: One of the following
```
Comment (Borda) [inline src/torchmetrics/wrappers/feature_share.py:54]:
```suggestion
        metrics: One of the following:
```

File: src/torchmetrics/wrappers/feature_share.py [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+            raise TypeError(f"max_cache_size should be an integer, but got {max_cache_size}")
+
+        try:
+            first_net = next(iter(self.values()))
+            network_to_share = getattr(first_net, first_net.feature_network)
+        except AttributeError as e:
+            raise AttributeError(
+                "Tried to extract the network to share from the first metric, but it did not have a `feature_network`"
+                " attribute. Please make sure that the metric has a attribute with that name, else it cannot be shared."
+            ) from e
```
Comment (Borda) [inline src/torchmetrics/wrappers/feature_share.py:104]:
```suggestion
        except AttributeError as err:
            raise AttributeError(
                "Tried to extract the network to share from the first metric, but it did not have a `feature_network`"
                " attribute. Please make sure that the metric has an attribute with that name, else it cannot be shared."
            ) from err
```

File: src/torchmetrics/wrappers/feature_share.py [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+        cached_net = NetworkCache(network_to_share, max_size=max_cache_size)
+
+        # set the cached network to all metrics
+        for metric_name, metric in self.items():
+            if not hasattr(metric, "feature_network"):
+                raise AttributeError(
+                    "Tried to set the cached network to all metrics, but one of the metrics did not have a "
+                    "`feature_network` attribute. Please make sure that all metrics have a attribute with that name, "
+                    f"else it cannot be shared. Failed on metric {metric_name}."
```
Comment (Borda) [inline src/torchmetrics/wrappers/feature_share.py:113]:
```suggestion
                    "Tried to set the cached network to all metrics, but one of the metrics did not have a"
                    " `feature_network` attribute. Please make sure that all metrics have a attribute with that name,"
                    f" else it cannot be shared. Failed on metric {metric_name}."
```

File: src/torchmetrics/wrappers/feature_share.py [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+                )
+
+            # check if its the same network as the first metric
+            if str(getattr(metric, metric.feature_network)) != str(network_to_share):
+                rank_zero_warn(
+                    f"The network to share between the metrics is not the same for all metrics. "
+                    f"Metric {metric_name} has a different network than the first metric."
+                    "This may lead to unexpected behavior.",
```
Comment (Borda) [inline src/torchmetrics/wrappers/feature_share.py:121]:
```suggestion
                    f"The network to share between the metrics is not the same for all metrics."
                    f" Metric {metric_name} has a different network than the first metric."
                    " This may lead to unexpected behavior.",
```

**Review by Borda — APPROVED** [review]

**Review by stancld — APPROVED** [review]
